File: web/src/shared/data/weapon-types.ts
/**
 * Weapon definition schema.
 *
 * Every number that affects time-to-kill lives in a WeaponDef, so balance changes
 * never require touching combat code. Attachments express themselves as a sparse
 * set of modifiers over these same fields (see attachments.ts), which means a
 * single `resolveWeapon()` pass produces the effective stats for any build.
 */

export enum WeaponClass {
  AssaultRifle = 'assault_rifle',
  SubmachineGun = 'smg',
  LightMachineGun = 'lmg',
  SniperRifle = 'sniper',
  MarksmanRifle = 'marksman',
  Shotgun = 'shotgun',
  Pistol = 'pistol',
  Launcher = 'launcher',
  Melee = 'melee',
  Special = 'special',
}

export enum FireMode {
  /** Holds trigger, keeps firing. */
  Auto = 'auto',
  /** One shot per trigger pull. */
  Semi = 'semi',
  /** Fixed-length burst per pull. */
  Burst = 'burst',
  /** Manual cycle between shots (bolt/pump). */
  BoltAction = 'bolt',
  /** Melee swing. */
  Swing = 'swing',
}

export enum AttachmentSlot {
  Muzzle = 0,
  Barrel = 1,
  Optic = 2,
  Underbarrel = 3,
  Magazine = 4,
  Stock = 5,
  RearGrip = 6,
  Laser = 7,
}

export const ATTACHMENT_SLOT_COUNT = 8;
/** Players may equip at most this many attachments, COD-style. */
export const MAX_EQUIPPED_ATTACHMENTS = 5;

/**
 * A damage curve is a list of (distance, damage) stops. Damage is linearly
 * interpolated between stops and clamped flat beyond the last one. This models
 * COD's stepped damage ranges while keeping the falloff smooth enough that
 * players don't feel a cliff edge.
 */
export interface DamageStop {
  /** Distance in metres at which this damage value applies. */
  distance: number;
  damage: number;
}

/**
 * Deterministic recoil. COD guns kick along a repeatable pattern so the gun can
 * be mastered; a small random component keeps it from being a laser.
 *
 * `pattern` is sampled by shot index (clamped at the end), giving per-shot
 * vertical/horizontal kick in radians.
 */
export interface RecoilProfile {
  /** Per-shot kick, indexed by shot number in the current trigger pull. */
  pattern: Array<{ pitch: number; yaw: number }>;
  /** Random kick added on top of the pattern, in radians. */
  randomPitch: number;
  randomYaw: number;
  /** How fast accumulated recoil decays back toward zero, per second. */
  recoverySpeed: number;
  /** Fraction of recoil the view returns automatically (1 = full auto-centering). */
  recoveryFraction: number;
  /** Visual-only kick applied to the viewmodel, multiplied against the above. */
  viewKickMultiplier: number;
  /** Camera shake magnitude in radians. */
  cameraShake: number;
}

/**
 * Bullet spread. Hipfire is a cone that grows while moving and firing; ADS
 * collapses it toward `adsMin`. Snipers reach exactly zero when fully scoped.
 */
export interface SpreadProfile {
  /** Cone half-angle in radians while standing still and hip-firing. */
  hipMin: number;
  hipMax: number;
  /** Fully-aimed cone half-angle. */
  adsMin: number;
  adsMax: number;
  /** Added per shot fired. */
  perShot: number;
  /** Decay back toward the minimum, per second. */
  recovery: number;
  /** Multipliers applied to the current cone. */
  movingMultiplier: number;
  jumpingMultiplier: number;
  crouchMultiplier: number;
  proneMultiplier: number;
}

export interface WeaponHandling {
  /** Seconds from hipfire to fully aimed. The single most important SMG/AR stat. */
  adsTime: number;
  /** Seconds from sprinting to being able to fire. */
  sprintOutTime: number;
  /** Seconds to raise this weapon when swapping to it. */
  drawTime: number;
  /** Seconds to lower this weapon when swapping away. */
  holsterTime: number;
  /** Tactical reload keeps the chambered round. */
  reloadTime: number;
  /** Empty reload has to cycle the action. */
  reloadEmptyTime: number;
  /** Point during the reload at which ammo is actually added (reload cancelling). */
  reloadAmmoTime: number;
  reloadEmptyAmmoTime: number;
  /** Multiplier on base movement speed while holding this weapon. */
  movementSpeedMultiplier: number;
  /** Multiplier on movement speed while aiming. */
  adsSpeedMultiplier: number;
  /** Weapon sway amplitude while aimed, radians. */
  swayAmount: number;
  swaySpeed: number;
}

export interface WeaponDef {
  id: string;
  name: string;
  /** Short label for the killfeed and HUD. */
  shortName: string;
  class: WeaponClass;
  description: string;

  fireMode: FireMode;
  /** Rounds per minute. For burst weapons this is the in-burst rate. */
  rpm: number;
  /** Shots per burst; ignored unless fireMode is Burst. */
  burstCount: number;
  /** Delay between bursts, seconds. */
  burstDelay: number;

  magSize: number;
  startingReserve: number;
  maxReserve: number;
  /** Pellets per shot. Shotguns fire many; everything else fires one. */
  pellets: number;

  damage: DamageStop[];
  /** Damage multiplier applied to non-player targets (killstreaks, deployables). */
  vehicleDamageMultiplier: number;
  /** How much material this round punches through, 0..1. Feeds wallbang math. */
  penetration: number;
  /** Muzzle velocity in m/s. Infinity = pure hitscan (most COD guns). */
  muzzleVelocity: number;
  /** Projectile drop in m/s². 0 for hitscan. */
  bulletGravity: number;

  recoil: RecoilProfile;
  spread: SpreadProfile;
  handling: WeaponHandling;

  /** Which attachment slots this weapon supports. */
  attachmentSlots: AttachmentSlot[];

  /** Player level required to unlock. 0 = available from the start. */
  unlockLevel: number;

  /** Audio synthesis parameters — see the procedural audio engine. */
  audio: {
    /** Base pitch of the shot's body, Hz. */
    bodyFreq: number;
    /** Duration of the transient crack, seconds. */
    crackDuration: number;
    /** Low-end thump amount, 0..1. */
    boom: number;
    /** Mechanical action noise, 0..1. */
    mech: number;
    /** Tail length (reverb send), seconds. */
    tail: number;
    /** True if this weapon is suppressed by default. */
    suppressed: boolean;
  };

  /** Viewmodel construction parameters — see the procedural model builder. */
  model: {
    /** Overall length in metres, drives the primitive layout. */
    length: number;
    /** Body colour. */
    color: number;
    accentColor: number;
    /** Magazine size/shape hint. */
    magStyle: 'stick' | 'box' | 'drum' | 'tube' | 'none';
    stockStyle: 'fixed' | 'folding' | 'skeleton' | 'none';
    hasCarryHandle: boolean;
    barrelLength: number;
    /** Position of the iron sights relative to the barrel, for ADS alignment. */
    sightHeight: number;
  };

  /** Zoom factor applied to FOV when aiming (1 = no zoom). */
  adsZoom: number;
  /** True for weapons that show a scope overlay instead of the viewmodel. */
  scoped: boolean;
  /** Fraction of a second the scope takes to focus (sniper glint / blur). */
  scopeFocusTime: number;

  /** Melee damage when bashing with this weapon. */
  meleeDamage: number;
  /** Extra flags that combat code branches on. */
  traits: WeaponTrait[];
}

export enum WeaponTrait {
  /** One-hit-kill to the upper torso within the first damage stop. */
  OneShotUpperTorso = 'one_shot_upper_torso',
  /** Fires explosive rounds. */
  Explosive = 'explosive',
  /** Cannot be dual-wielded or akimbo'd. */
  NoAkimbo = 'no_akimbo',
  /** Rechambers between shots, blocking fire. */
  Rechamber = 'rechamber',
  /** Reloads one round at a time and can be interrupted. */
  ShellReload = 'shell_reload',
  /** Ignores the ADS sway penalty. */
  SteadyAim = 'steady_aim',
  /** Locks onto air targets. */
  AirLockOn = 'air_lock_on',
  /** Damage is unaffected by range. */
  NoFalloff = 'no_falloff',
  /** Silent by design — never shows on the minimap. */
  AlwaysSuppressed = 'always_suppressed',
}

// ---------------------------------------------------------------------------
// Derived helpers
// ---------------------------------------------------------------------------

/** Seconds between shots, derived from RPM. */
export function fireInterval(def: Pick<WeaponDef, 'rpm'>): number {
  return def.rpm > 0 ? 60 / def.rpm : 0;
}

/**
 * Damage at a given distance, interpolating between stops.
 * Flat before the first stop and after the last, which is what makes the
 * "damage range" concept legible to players.
 */
export function damageAtRange(stops: readonly DamageStop[], distance: number): number {
  if (stops.length === 0) return 0;
  const first = stops[0]!;
  if (distance <= first.distance) return first.damage;

  for (let i = 1; i < stops.length; i++) {
    const prev = stops[i - 1]!;
    const cur = stops[i]!;
    if (distance <= cur.distance) {
      const span = cur.distance - prev.distance;
      if (span <= 0) return cur.damage;
      const t = (distance - prev.distance) / span;
      return prev.damage + (cur.damage - prev.damage) * t;
    }
  }
  return stops[stops.length - 1]!.damage;
}

/**
 * Shots required to kill at a given range and hit multiplier.
 * The canonical way to reason about balance — every weapon in the arsenal is
 * validated against expected STK bands in the test suite.
 */
export function shotsToKill(
  stops: readonly DamageStop[],
  distance: number,
  health: number,
  multiplier = 1,
): number {
  const dmg = damageAtRange(stops, distance) * multiplier;
  if (dmg <= 0) return Infinity;
  return Math.ceil(health / dmg);
}

/** Time to kill in seconds, ignoring the first shot's travel time. */
export function timeToKill(
  def: Pick<WeaponDef, 'rpm' | 'damage' | 'pellets' | 'burstCount' | 'burstDelay' | 'fireMode'>,
  distance: number,
  health: number,
  multiplier = 1,
): number {
  const perShot = damageAtRange(def.damage, distance) * multiplier * Math.max(1, def.pellets);
  if (perShot <= 0) return Infinity;
  const shots = Math.ceil(health / perShot);
  if (shots <= 1) return 0;

  const interval = fireInterval(def);
  if (def.fireMode === FireMode.Burst && def.burstCount > 1) {
    const fullBursts = Math.floor((shots - 1) / def.burstCount);
    const remainder = (shots - 1) % def.burstCount;
    return fullBursts * ((def.burstCount - 1) * interval + def.burstDelay) + remainder * interval;
  }
  return (shots - 1) * interval;
}

File: dotnet/src/OperationVanguard.Core/Data/WeaponTypes.cs
using System.Text.Json.Serialization;

namespace OperationVanguard.Core;

[JsonConverter(typeof(JsonStringEnumConverter<WeaponClass>))]
public enum WeaponClass
{
    [JsonStringEnumMemberName("assault_rifle")] AssaultRifle,
    [JsonStringEnumMemberName("smg")] SubmachineGun,
    [JsonStringEnumMemberName("lmg")] LightMachineGun,
    [JsonStringEnumMemberName("sniper")] SniperRifle,
    [JsonStringEnumMemberName("marksman")] MarksmanRifle,
    [JsonStringEnumMemberName("shotgun")] Shotgun,
    [JsonStringEnumMemberName("pistol")] Pistol,
    [JsonStringEnumMemberName("launcher")] Launcher,
    [JsonStringEnumMemberName("melee")] Melee,
    [JsonStringEnumMemberName("special")] Special,
}

[JsonConverter(typeof(JsonStringEnumConverter<FireMode>))]
public enum FireMode
{
    [JsonStringEnumMemberName("auto")] Auto,
    [JsonStringEnumMemberName("semi")] Semi,
    [JsonStringEnumMemberName("burst")] Burst,
    [JsonStringEnumMemberName("bolt")] BoltAction,
    [JsonStringEnumMemberName("swing")] Swing,
}

public enum AttachmentSlot
{
    Muzzle = 0,
    Barrel = 1,
    Optic = 2,
    Underbarrel = 3,
    Magazine = 4,
    Stock = 5,
    RearGrip = 6,
    Laser = 7,
}

[JsonConverter(typeof(JsonStringEnumConverter<WeaponTrait>))]
public enum WeaponTrait
{
    [JsonStringEnumMemberName("one_shot_upper_torso")] OneShotUpperTorso,
    [JsonStringEnumMemberName("explosive")] Explosive,
    [JsonStringEnumMemberName("no_akimbo")] NoAkimbo,
    [JsonStringEnumMemberName("rechamber")] Rechamber,
    [JsonStringEnumMemberName("shell_reload")] ShellReload,
    [JsonStringEnumMemberName("steady_aim")] SteadyAim,
    [JsonStringEnumMemberName("air_lock_on")] AirLockOn,
    [JsonStringEnumMemberName("no_falloff")] NoFalloff,
    [JsonStringEnumMemberName("always_suppressed")] AlwaysSuppressed,
}

public sealed class DamageStop
{
    public double Distance { get; set; }
    public double Damage { get; set; }

    public DamageStop Clone() => new() { Distance = Distance, Damage = Damage };
}

public sealed class RecoilStep
{
    public double Pitch { get; set; }
    public double Yaw { get; set; }

    public RecoilStep Clone() => new() { Pitch = Pitch, Yaw = Yaw };
}

public sealed class RecoilProfile
{
    public List<RecoilStep> Pattern { get; set; } = [];
    public double RandomPitch { get; set; }
    public double RandomYaw { get; set; }
    public double RecoverySpeed { get; set; }
    public double RecoveryFraction { get; set; }
    public double ViewKickMultiplier { get; set; }
    public double CameraShake { get; set; }

    public RecoilProfile Clone() => new()
    {
        Pattern = Pattern.Select(step => step.Clone()).ToList(),
        RandomPitch = RandomPitch,
        RandomYaw = RandomYaw,
        RecoverySpeed = RecoverySpeed,
        RecoveryFraction = RecoveryFraction,
        ViewKickMultiplier = ViewKickMultiplier,
        CameraShake = CameraShake,
    };
}

public sealed class SpreadProfile
{
    public double HipMin { get; set; }
    public double HipMax { get; set; }
    public double AdsMin { get; set; }
    public double AdsMax { get; set; }
    public double PerShot { get; set; }
    public double Recovery { get; set; }
    public double MovingMultiplier { get; set; }
    public double JumpingMultiplier { get; set; }
    public double CrouchMultiplier { get; set; }
    public double ProneMultiplier { get; set; }

    public SpreadProfile Clone() => (SpreadProfile)MemberwiseClone();
}

public sealed class WeaponHandling
{
    public double AdsTime { get; set; }
    public double SprintOutTime { get; set; }
    public double DrawTime { get; set; }
    public double HolsterTime { get; set; }
    public double ReloadTime { get; set; }
    public double ReloadEmptyTime { get; set; }
    public double ReloadAmmoTime { get; set; }
    public double ReloadEmptyAmmoTime { get; set; }
    public double MovementSpeedMultiplier { get; set; }
    public double AdsSpeedMultiplier { get; set; }
    public double SwayAmount { get; set; }
    public double SwaySpeed { get; set; }

    public WeaponHandling Clone() => (WeaponHandling)MemberwiseClone();
}

public sealed class WeaponAudioDef
{
    public double BodyFreq { get; set; }
    public double CrackDuration { get; set; }
    public double Boom { get; set; }
    public double Mech { get; set; }
    public double Tail { get; set; }
    public bool Suppressed { get; set; }

    public WeaponAudioDef Clone() => (WeaponAudioDef)MemberwiseClone();
}

public sealed class WeaponModelDef
{
    public double Length { get; set; }
    public int Color { get; set; }
    public int AccentColor { get; set; }
    public string MagStyle { get; set; } = "none";
    public string StockStyle { get; set; } = "none";
    public bool HasCarryHandle { get; set; }
    public double BarrelLength { get; set; }
    public double SightHeight { get; set; }

    public WeaponModelDef Clone() => (WeaponModelDef)MemberwiseClone();
}

public sealed class WeaponDef
{
    public string Id { get; set; } = "";
    public string Name { get; set; } = "";
    public string ShortName { get; set; } = "";
    public WeaponClass Class { get; set; }
    public string Description { get; set; } = "";
    public FireMode FireMode { get; set; }
    public double Rpm { get; set; }
    public int BurstCount { get; set; }
    public double BurstDelay { get; set; }
    public int MagSize { get; set; }
    public int StartingReserve { get; set; }
    public int MaxReserve { get; set; }
    public int Pellets { get; set; }
    public List<DamageStop> Damage { get; set; } = [];
    public double VehicleDamageMultiplier { get; set; }
    public double Penetration { get; set; }
    public double MuzzleVelocity { get; set; }
    public double BulletGravity { get; set; }
    public RecoilProfile Recoil { get; set; } = new();
    public SpreadProfile Spread { get; set; } = new();
    public WeaponHandling Handling { get; set; } = new();
    public List<AttachmentSlot> AttachmentSlots { get; set; } = [];
    public int UnlockLevel { get; set; }
    public WeaponAudioDef Audio { get; set; } = new();
    public WeaponModelDef Model { get; set; } = new();
    public double AdsZoom { get; set; }
    public bool Scoped { get; set; }
    public double ScopeFocusTime { get; set; }
    public double MeleeDamage { get; set; }
    public List<WeaponTrait> Traits { get; set; } = [];

    public WeaponDef Clone() => new()
    {
        Id = Id,
        Name = Name,
        ShortName = ShortName,
        Class = Class,
        Description = Description,
        FireMode = FireMode,
        Rpm = Rpm,
        BurstCount = BurstCount,
        BurstDelay = BurstDelay,
        MagSize = MagSize,
        StartingReserve = StartingReserve,
        MaxReserve = MaxReserve,
        Pellets = Pellets,
        Damage = Damage.Select(stop => stop.Clone()).ToList(),
        VehicleDamageMultiplier = VehicleDamageMultiplier,
        Penetration = Penetration,
        MuzzleVelocity = MuzzleVelocity,
        BulletGravity = BulletGravity,
        Recoil = Recoil.Clone(),
        Spread = Spread.Clone(),
        Handling = Handling.Clone(),
        AttachmentSlots = [.. AttachmentSlots],
        UnlockLevel = UnlockLevel,
        Audio = Audio.Clone(),
        Model = Model.Clone(),
        AdsZoom = AdsZoom,
        Scoped = Scoped,
        ScopeFocusTime = ScopeFocusTime,
        MeleeDamage = MeleeDamage,
        Traits = [.. Traits],
    };
}

public static class WeaponMath
{
    public static double FireInterval(WeaponDef def) => def.Rpm > 0 ? 60 / def.Rpm : 0;

    public static double DamageAtRange(IReadOnlyList<DamageStop> stops, double distance)
    {
        if (stops.Count == 0) return 0;
        var first = stops[0];
        if (distance <= first.Distance) return first.Damage;

        for (var i = 1; i < stops.Count; i++)
        {
            var previous = stops[i - 1];
            var current = stops[i];
            if (distance > current.Distance) continue;
            var span = current.Distance - previous.Distance;
            if (span <= 0) return current.Damage;
            var t = (distance - previous.Distance) / span;
            return previous.Damage + (current.Damage - previous.Damage) * t;
        }

        return stops[^1].Damage;
    }

    public static double ShotsToKill(
        IReadOnlyList<DamageStop> stops,
        double distance,
        double health,
        double multiplier = 1)
    {
        var damage = DamageAtRange(stops, distance) * multiplier;
        return damage <= 0 ? double.PositiveInfinity : Math.Ceiling(health / damage);
    }

    public static double TimeToKill(WeaponDef def, double distance, double health, double multiplier = 1)
    {
        var perShot = DamageAtRange(def.Damage, distance) * multiplier * Math.Max(1, def.Pellets);
        if (perShot <= 0) return double.PositiveInfinity;
        var shots = (int)Math.Ceiling(health / perShot);
        if (shots <= 1) return 0;

        var interval = FireInterval(def);
        if (def.FireMode == FireMode.Burst && def.BurstCount > 1)
        {
            var fullBursts = (shots - 1) / def.BurstCount;
            var remainder = (shots - 1) % def.BurstCount;
            return fullBursts * ((def.BurstCount - 1) * interval + def.BurstDelay)
                   + remainder * interval;
        }
        return (shots - 1) * interval;
    }
}
